xcube_cmems/cmems.py

```diff
@@ -97,40 +101,69 @@ def get_opendap_urls(self, data_id) -> List[str]:
-            startposition=rec + 1,
-            maxrecords=50,
-            sortby=sortby,
-            esn='full')
-        for record in csw.records.values():
-            if len(record.uris) > 0:
-                for uris in record.uris:
-                    if uris['protocol'] == 'WWW:OPENDAP':
-                        if uris['url']:
-                            opendap_uri = uris['url']
-                            scheme, netloc, path, query, fragment = \
-                                urlsplit(opendap_uri)
-                            split_paths = path.split('/')
-                            self.opendap_dataset_ids[split_paths[-1]] = \
-                                record.title
-
-    async def get_csw_records_concurrently(self, csw) -> None:
+    @staticmethod
+    async def get_response(session: aiohttp.ClientSession,
+                           url: str,
+                           params: Dict) -> \
+            Optional[aiohttp.ClientResponse]:
+        num_retries = 10
+        for i in range(num_retries):
+            resp = await session.request(method='GET',
+                                         url=url,
+                                         ssl=True,
+                                         params=params)
+            if resp.status == 200:
+                return resp
+            elif resp.status == 429:
+                time.sleep(10)
+            else:
+                break
+        return None
+
+    async def read_data_ids_from_csw_records(self, rec, max_records, session):
+        params = {
+            'service': 'CSW',
+            'request': 'GetRecords',
+            'version': '2.0.2',
+            'resultType': 'results',
+            'ElementSetName': 'full',
+            'startPosition': rec + 1,
+            'maxRecords': max_records
+        }
+        resp = await self.get_response(
+            session, self._csw_url, params
+        )
+        if not resp:
+            return
+        records_xml = etree.parse(BytesIO(await resp.content.read()))
+        opendap_elements = records_xml.getroot().findall(
+            './csw:SearchResults/csw:Record/dc:URI[@protocol="WWW:OPENDAP"]',
+            namespaces=CSW_NAMESPACES
+        )
+        for opendap_element in opendap_elements:
+            name = opendap_element.get('name')
+            title = opendap_element.getparent().find(
+                'dc:title',
+                namespaces=CSW_NAMESPACES
+            ).text
+            self.opendap_dataset_ids[name] = title
+
+    async def read_data_ids(self) -> None:
         """
         get csw records concurrently
         """
         tasks = []
-        num_of_records = 50
-        total_records = 300
-        for rec in range(0, total_records, num_of_records):
-            task = asyncio.ensure_future(self.get_record_from_csw(csw, rec))
-            tasks.append(task)
-        await asyncio.gather(*tasks)
+        records_per_request = 25
+        total_num_of_records = 300
+
```

Comment:
can we declare these as constants in .constants ?

Reply:
> I am not too happy with having the total number of records as a constant

Bad design. This will sooner or later fall on our feet.
I don't even understand, why it must be a constant, e.g., make this function a generator.

Reply:
It would be possible to turn this into a gnerator function, but that would come with a performance loss, as requests are handled sequentially again. I think what could solve this issue is the use of [Task Groups](https://docs.python.org/3/library/asyncio-task.html#task-groups), which do not require that the amount of tasks is known at the point the computation starts, but that requires Python 3.11. Shall I go for that?

Reply:
I would vote to include these changes in the next release and work on the possibility to use Task Groups in a different PR with a dedicated issue. The reason being the performance advantage of get_data_ids included in this PR.